woocommerce/woocommerce-ios
# [Mobile Payments] Printing receipts. Mark I: Hardware

Closes #3977 

⚠️ This PR is against the feature branch implementing support for Card Present Payments, not against develop ⚠️
⚠️ Testing this PR requires a hardware reader ⚠️

We continue building our prototype (for reference, p91TBi-4q4-p2), this time, by adding the scaffolding for printing receipts to AirPrint.

This PR adds a new interface to Hardware, and a barebones implementation of that interface that integrates with AirPrint. It also adds a couple of new model objects to wrap all the information we need to add to the receipt.

For more context about what needs to go into a receipt, please check [Stripe's documentation](https://stripe.com/docs/terminal/checkout/receipts#custom).


https://user-images.githubusercontent.com/2722505/115439566-b666ce80-a1dc-11eb-9008-5d1dcf6124b6.mov



## Changes
* Add ReceiptPrinterService. Expose this service via the ServiceLocator.
* Add an implementation of ReceiptPrinterService that prints using AirPrint. This class will be iterated on while working on issues #3982 and #3978
* Add a barebones implementation of a subclass of UIPrintPageRenderer that will be responsible for actually render the content of the receipt. To be fully implemented in #3978 
* Add a couple of model objects to wrap the information we need to pass to the ReceiptPrinterService.

## How to test
Testing is a bit complicated because:

1. Requires setting up a store for Card Present Payment Testing (also P91TBi-4BH-p2 for more specific instructions)
2. Requires an external hardware reader.

* Checkout the branch, run bundle exec pod install, just for kicks and giggles.
* Build and run on a device.
* Switch an external card reader on. The only card reader supported by now is the BBPOS Chipper 2X BT.
* Log in to an account that matches a store that has been configured for testing.
* Navigate to Settings in the app. (Tap the gear button in the top right)
* In the "Store Settings" section, tap "Manage Card Reader".
* Make sure the card reader is on and the blue light is blinking (we don't do any error handling yet, so it's important that things are setup for the happy path)
* Tap "Connect card reader"
* Wait until the app suggests one reader to connect to.
* Tap "Connect to reader"
* Make sure the reader light turns solid blue.
* Navigate to an order eligible for Card Present Payments (it has to be a store enrolled to WCPay, and the order needs to have as payment method associated Cash on Delivery or none).
* Tap the "Collect Payment" button.
* Tap the card to the reader when requested.
* If everything goes well, the card should be charged, and the payment should be completed.
* At that point, and only in order to test this PR, a preview of the receipt should be presented. At this point, the preview only shows a page number and the actual amount and currency of the order that was just charged using the card reader. The amount and currency should match the Order's.

Note: this is not the actual flow that will be implemented in the final beta. There is work pending in tickets #3979, #3982, #3981 #3978




Update release notes:
